design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+    1.  Cluster health and object/file functionality will be verified before the upgrade controller moves on to the next instances.
+
+### **Upgrade Health Verification**
+As mentioned previously, the manual health verification steps found in the [upgrade user guide](../Documentation/upgrade.md#health-verification) will be used by the upgrade controller, in an automated fashion, to ensure the cluster is healthy before proceeding with the upgrade process.
+This approach of upgrading one component, verifying health and stability, then upgrading the next component can be viewed as a form of [canary deployment](https://kubernetes.io/docs/concepts/cluster-administration/manage-deployment/#canary-deployments).
+
+Here is a quick summary of the standard health checks the upgrade controller should perform:
+* All pods are in the `Running` state and have few, if any, restarts
+  * No pods enter a crash loop backoff
+* Overall status: The overall cluster status is `OK` and there are no warning or error status messages displayed.
+* Monitors:  All of the monitors are `in quorum` and have individual status of `OK`.
+* OSDs: All OSDs are `UP` and `IN`.
+* MGRs: All Ceph managers are in the `Active` state.
+* Placement groups: All PGs are in the `active+clean` state.
+
+#### Pod Readiness/Liveness Probes
+To further supplement the upgrade controller's ability to determine health, as well as facilitate the built-in Kubernetes upgrade capabilities, the Rook pods should implement [liveness and readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/) when possible.
+For pods that implement these probes, the upgrade controller can check them as another data point in determining if things are healthy before proceeding with the upgrade.
+
+### **Rollback**
+If the upgrade controller observes the cluster to be in an unhealthy state (that does not recover) during the upgrade process, it will need to roll back components in the cluster to the previous stable version.
+This is possible due to the rolling/canary approach of the upgrade controller.
+To roll a component back to the previous version, the controller will simply set the `image` field of the pod template spec to the previous version then terminate each pod to allow their managing controller to start a new pod on the *old* version to replace it.
+
+The hope is that cluster health and stability will be restored once it has been rolled back to the previous version, but it is possible that simply rolling back the version may not solve all cases of cluster instability that begin during an upgrade process.
+We will need more hands on experience with cluster upgrades in order to improve both upgrade reliability and rollback effectiveness.
+
+### **Experience**
+#### Automatic Upgrades
+As described earlier, the cluster administrator manually begins the upgrade process by editing the cluster CRD to indicate a new software version is desired.
```

Comment:
i am not sure if this has a real use case. Admin users dont tent do this specially on live/production deployment. It is kinda scary for them for their deployment to upgrade automatically on some random time.

Reply:
you're probably right.  kubernetes doesn't have any automatic upgrade feature either.  i'll remove this from the design for now.